File: HowTheyWrite_android/app/src/main/java/com/kaiserpudding/howtheywrite/repositories/LessonRepository.java
package com.kaiserpudding.howtheywrite.repositories;

import android.app.Application;
import android.arch.lifecycle.LiveData;
import com.kaiserpudding.howtheywrite.database.AppDatabase;
import com.kaiserpudding.howtheywrite.database.dao.LessonCharacterJoinDao;
import com.kaiserpudding.howtheywrite.database.dao.LessonDao;
import com.kaiserpudding.howtheywrite.model.Character;
import com.kaiserpudding.howtheywrite.model.Lesson;
import com.kaiserpudding.howtheywrite.model.LessonCharacterJoin;
import java.util.List;

public class LessonRepository {

  private LessonDao lessonDao;
  private LessonCharacterJoinDao lessonCharacterJoinDao;

  public LessonRepository(Application application) {
    AppDatabase db = AppDatabase.getDatabase(application);
    lessonDao = db.lessonDao();
    lessonCharacterJoinDao = db.lessonCharacterJoinJoinDao();
  }

  public void insert(Lesson lesson) {
    lessonDao.insert(lesson);
    lessonDao.insert(lesson);
    insertLessonCharacterJoin(lesson);
  }

  public void insertLessonCharacterJoin(Lesson lesson) {
    List<Character> characters = lesson.getCharacters();
    int numberOfCharacters = characters.size();
    if (numberOfCharacters > 0) {
      int lessonId = lesson.getId();
      LessonCharacterJoin[] toInsert = new LessonCharacterJoin[numberOfCharacters];
      for (int i = 0; i < numberOfCharacters; i++) {
        toInsert[i] = new LessonCharacterJoin(lessonId, characters.get(i).getId());
      }
      lessonCharacterJoinDao.insertLessonCharacterJoin(toInsert);
    }
  }

  public LiveData<List<Lesson>> getAllLiveDataLessons() {
    return lessonDao.getAllLessons();
  }

  public Lesson getLessonWithRelationById(int id) {
    return lessonCharacterJoinDao.getLessonByLessonId(id);
  }

  public LiveData<Lesson> getLiveDataLessonWithRelationById(int id) {
    return lessonCharacterJoinDao.getLiveDataLessonByLessonId(id);
  }

  public List<String> getAllLessonNames() {
    return lessonDao.getAllLessonNames();
  }
}

File: HowTheyWrite_android/app/src/main/java/com/kaiserpudding/howtheywrite/repositories/LessonRepository.kt
package com.kaiserpudding.howtheywrite.repositories

import android.app.Application
import androidx.lifecycle.LiveData
import com.kaiserpudding.howtheywrite.database.AppDatabase
import com.kaiserpudding.howtheywrite.database.dao.LessonCharacterJoinDao
import com.kaiserpudding.howtheywrite.database.dao.LessonDao
import com.kaiserpudding.howtheywrite.model.Lesson
import com.kaiserpudding.howtheywrite.model.LessonCharacterJoin
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class LessonRepository(application: Application) {

    private val lessonDao: LessonDao
    private val lessonCharacterJoinDao: LessonCharacterJoinDao


    init {
        val db = AppDatabase.getDatabase(application)
        lessonDao = db.lessonDao()
        lessonCharacterJoinDao = db.lessonCharacterJoinJoinDao()
    }

    suspend fun insert(lesson: Lesson) {
        withContext(Dispatchers.IO) {
            lessonDao.insert(lesson)
            insertLessonCharacterJoin(lesson)
        }

    }

//    suspend fun delete(lesson: Lesson) {
//        withContext(Dispatchers.IO) {
//            lessonDao.delete(lesson)
//            //TODO delete lessonCharacterJoin
//        }
//    }

    fun allLiveDataLessons(): LiveData<List<Lesson>> {
        return lessonDao.allLessons()
    }

    private suspend fun insertLessonCharacterJoin(lesson: Lesson) {
        withContext(Dispatchers.IO) {
            val characters = lesson.characters
            val numberOfCharacters = characters.size
            if (numberOfCharacters > 0) {
                val lessonId = lesson.id
                val toInsert = arrayOf<LessonCharacterJoin>()
                for (i in 0 until numberOfCharacters) {
                    toInsert[i] = LessonCharacterJoin(lessonId, characters[i].id)
                }
                lessonCharacterJoinDao.insertLessonCharacterJoin(toInsert)
            }
        }
    }

    fun getLiveDataLessonNameById(lessonId: Long): LiveData<String> {
        return lessonDao.getLiveDataLessonNameById(lessonId)
    }

    suspend fun deleteLessons(lessonIds: LongArray) {
        withContext(Dispatchers.IO) {
            lessonDao.deleteLessons(lessonIds)
            lessonCharacterJoinDao.deleteAllLessonCharacterJoinsFromLesson(lessonIds)
        }
    }
}
